README.txt
README
------

Data Basin
Data Access Tool for Salesforce

API implementation is inside DataBasinKit


Current Version: 1.1 S
----------------------

NEWS
- improved interface with ComboBoxes and preferences to filter out more system objects
- remember last succesful login username
- getUpdated / getDeleted
- Undelete
- Improved support for sub-objects and lists and unpacking (from DataBasinKit)
- support for enabling Assignment Rules in create/update


Older news:
- multi-threading: concurrent operations with interrupt
- down batch size
- handle multiple errors in update or create
- filter line-breaks when writing CSV
- Query gets parsed and this allows to write fields in Query Order
- Query parsing allows for fields to be empty and still preserve the correct number of columns in the CSV file
- Download batch size settable
- CVS qualifier and separator settable in both reading and writing
- Support LIMIT clause in Select Identify
- Improved Exception catching and display for Select Identify
- Object Inspector supports selection of items
- Support count and aggregate queries
- Select Identify works with multiple columns (all in AND as condition)
- Support "." as Semi-Join
- Query Identify: run a select on each item in the list, the single column provided in the file will be used as identifier and it must point to an Id or a unique identifier
- improved exception trapping and display
- support for reading quoted CSV files


LIMITATIONS
- CVS Reading and Writing does not support full quoting properly (but it improved vastly in 0.5)
- no field matching: create column names must match exactly salesforce field names

MISSING FEATURES
- Upsert


ChangeLog
2023-05-23 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m:
	Interpret entered text according to describe for DBSFDataTypes, thus
	allowing finally edit for formatted Date/DateTime

2021-04-18 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Respect assignment rules also in Object Inspector update.

2021-04-17 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.m
	Set runAssignmentRules for update and insert.

2020-08-21 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.m
	* Resources/DataBasin.gorm
	Use a ComboBox instead of PopUpButton for object list.

2020-05-21 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Remember last successful login username.

2020-05-20 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Filter specfic objects selected in preferences also in DescribeObject
	method.

2020-05-19 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	* Preferences.h
	* Preferences.m
	* Resources/Preferences.gorm
	Preferences to hide also ChangeEvent and Feed from object list.

2020-01-26 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Do not instantiate a new DBSoap, since it is now done in the _delete method itself

2019-06-02 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Update to new API using NSURL instead of NSString as service URL type.

2019-05-28 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.m
	* Resources/DataBasin.gorm
	Allow setting Session Info directly without login (useful for SSO)

2018-07-23 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.h
	* Resources/DataBasin.gorm
	getUpdated and undelete

2018-07-17 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.h
	* Resources/DataBasin.gorm
	getDeleted

2018-05-04 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.h
	* Resources/DataBasin.gorm
	Retrieve

2017-11-10 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Better exception handling

2017-10-31 Riccardo Mottola <rm@gnu.org>

	* AppController.m
        Allow HTML and XLS.

2017-10-31 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Add missing preference settings and improve reloadDefaults.

2017-10-26 Riccardo Mottola <rm@gnu.org>

	* Preferences.h
	* Preferences.m
	* Resources/Preferences.gorm
	New pref settings.

2017-10-17 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Handle DB SFTypes

2017-10-10 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Display NSNumbers

2017-10-04 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Set String Encoding for writers.

2017-10-03 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Do not call writeStart/writeEnd on FileWriter since DBSoapCSV does it

2017-09-26 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	enable XLS also for Select Identify

2017-09-25 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Call writeStart/writeEnd to correctly initialize the file.

2017-09-20 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	For select, start using different writers.

2017-04-26 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Set SoapHander on login, not on opening of the Object Inspector.

2017-02-21 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Fault panel must be shown in separate thread.

2016-12-15 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Do not leak on fault, display fault panel only through separate
	showException:

2016-11-23 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Fix wrong size in popup selection.

2016-07-04 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
        Load and update in separate thread.

2016-06-27 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
        Enable search menu.

2016-06-27 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
        Re-filter after sorting.

2016-06-24 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.h
	* DBObjectInspector.m
        * English.lproj/ObjectInspector.nib
        Allow filtering of columns by content - currently by developer name

2016-06-22 Riccardo Mottola <rm@gnu.org>

	* Preferences.h
	* Preferences.m
	* AppController.m
	* Resources/Preferences.gorm
	Introduce preference tu filter out Share and History objects and use it in Insert and Update PopUps

2016-06-17 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Visually mark updateable and updated cells.

2016-06-17 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
        Make cells editable/noneditable for non-updateable fiels.

2016-06-15 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.m
	Skip udpate list if value did not change, skip non-updateable fields.

2016-06-13 Riccardo Mottola <rm@gnu.org>

	* DBObjectInspector.h
        * DBObjectInspector.m
        * English.lproj/ObjectInspector.nib
        Edit values and Update from object inspector.

2016-04-19 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.m
	Set batch sizes when instantiating DBSoap.

2016-03-16 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Reset UI after insert completion

2015-12-11 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Fix Preferences defaults, do not set them wrongly on panel show.

2015-09-03 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Separate exception panel show in separate method and always call it on main thread (or sometimes old exception was shown).

2015-06-04 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Expose alert panels from main thread.

2015-06-04 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Reset gui  if an file creation error occours.

2015-04-28 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	* English.lproj/MainMenu.nib/keyedobjects.nib
	Add 100 and 200 sizes.

2015-04-28 Riccardo Mottola <rm@gnu.org>

	* AppController.m
	Pools need to be drained, not released, since they don't get retained.

2015-04-23 Riccardo Mottola <rm@gnu.org>

	* AppController.h
	* AppController.m
	Stop support in delete.

2015-04-23 Riccardo Mottola <rm@gnu.org>


DBSObject.m
/*
 Project: DataBasin
 DBSobject.m
 
 Copyright (C) 2010-2014 Free Software Foundation
 
 Author: Riccardo Mottola
 
 Created by Riccardo Mottola on 20/07/10.
 

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Library General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free
 Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 Boston, MA 02111 USA.
*/


#import "DBSObject.h"
#import "DBSoap.h"

@implementation DBSObject

/** <p>Returns the corresponding 18-character case-insensitive
    salesforce Id given the 15-character version <i>id15</i>
*/
+ (NSString *)idTo18: (NSString *) id15
{
  NSMutableString *suffix;
  int i;
  int j;
  
  if ([id15 length] != 15)
    return nil;
  
  suffix = [NSMutableString stringWithCapacity: 3];
  for (i = 0; i < 3; i++)
    {
      int flags = 0;
      for (j = 0; j < 5; j++)
        {
          unichar c;
          c = [id15 characterAtIndex: i * 5 + j];
          if (c >= 'A' && c <= 'Z')
            flags += 1 << j;
        }
      if (flags <= 25)
        [suffix appendString: [@"ABCDEFGHIJKLMNOPQRSTUVWXYZ" substringWithRange: NSMakeRange(flags, 1)]];
      else
        [suffix appendString: [@"012345" substringWithRange: NSMakeRange(flags-26, 1)]];
    }
  return [id15 stringByAppendingString: suffix];  
}

- (id)init
{
  if ((self = [super init]))
    {
      fieldNames = [[NSMutableArray arrayWithCapacity: 1] retain];
      fieldProperties = [[NSMutableDictionary dictionaryWithCapacity: 1] retain];
      recordValues = [[NSMutableDictionary dictionaryWithCapacity: 1] retain];
      objectProperties = [[NSMutableDictionary dictionaryWithCapacity: 1] retain];
    }
  return self;
}

- (id)copyWithZone: (NSZone *)zone
{
  DBSObject *sobj;

  sobj = [[[self class] allocWithZone: zone] init];
  if(sobj)
    {
      sobj->dbs = dbs;
      sobj->fieldNames = [fieldNames mutableCopy];
      sobj->fieldProperties = [fieldProperties mutableCopy];
      sobj->recordValues = [recordValues mutableCopy];
      sobj->objectProperties = [objectProperties mutableCopy];
      sobj->recordTypes = [recordTypes copy];
    }
  return sobj;
}

- (void)dealloc
{
  [fieldNames release];
  [fieldProperties release];
  [recordValues release];
  [objectProperties release];
  [recordTypes release];
  [super dealloc];
}

- (NSString *)description
{
  return [NSString stringWithFormat: @"<DBSObject %@, %@, %@>", [self name], [self sfId], [self label]];
}

/** <p>returns the current salesforce id, in whichever format it is currently stored.</p>
*/
- (NSString *)sfId
{
  return [recordValues objectForKey: @"Id"];
}

/** <p>Returns the current salesforce Id, always in the 15-character case-sensitive format,
    converting it if necessary.</p>
*/
- (NSString *)sfId15
{
  NSString *sfid;
  
  sfid = [recordValues objectForKey: @"Id"];
  if ([sfid length] == 18)
    sfid = [sfid substringToIndex: 15];
  
  if ([sfid length] != 15)
    {
      NSLog(@"Invalid ID: %@", sfid);
      return nil;
    }
  
  return sfid;
}

/** <p>Returns the current salesforce Id, always in the 18-character case-insensitive format,
converting it if necessary.</p>
*/
- (NSString *)sfId18
{
  NSString *sfid;
  
  sfid = [recordValues objectForKey: @"Id"];
  if ([sfid length] == 15)
    sfid = [DBSObject idTo18: sfid];
  
  if ([sfid length] != 18)
    {
      NSLog(@"Invalid ID: %@", sfid);
      return nil;
    }
  
  return sfid;
}

- (void)setDBSoap: (DBSoap *)db
{
  dbs = db;
}

- (void)setObjectProperties: (NSDictionary *)properties
{
  [objectProperties release];
  objectProperties = [[NSMutableDictionary dictionaryWithDictionary: properties] retain];
}

- (NSDictionary *)objectProperties
{
  return [NSDictionary dictionaryWithDictionary: objectProperties];
}

- (NSString *)name
{
  /*
  NSString *name;

  name = nil;
  if ([[self type] isEqualToString:@"Contract"])
    name = [recordValues objectForKey: @"ContractNumber"];
  else if ([[self type] isEqualToString:@"Case"])
    name = [recordValues objectForKey: @"CaseNumber"];
  else
    name = [recordValues objectForKey: @"Name"];

  if (name == nil)
    name = [objectProperties objectForKey: @"name"];

  return name;
  */
  return [objectProperties objectForKey: @"name"];
}

- (NSString *)type
{
  return [objectProperties objectForKey: @"type"];
}

- (NSString *)label
{
  return [objectProperties objectForKey: @"label"];
}

- (NSString *)keyPrefix
{
  return [objectProperties objectForKey: @"keyPrefix"];
}


- (void)setProperties: (NSDictionary *)properties forField: (NSString *)field
{
  // TODO should check if field is not already present
  [fieldNames addObject: field];
  [fieldProperties setObject: properties forKey: field];
}

- (void)removeField: (NSString *)field
{
  [recordValues removeObjectForKey: field];
  [fieldNames removeObject: field];
}

- (NSDictionary *)propertiesOfField: (NSString *)field
{
  return [fieldProperties objectForKey: field];
}

/** <p>Returns a list of all known field names.</p>
*/
- (NSArray *)fieldNames
{
  return fieldNames;
}

/** <p>Returns the value of field <i>field</i><p>
*/
- (id)valueForField: (NSString *)field
{
  return [recordValues objectForKey: field];
}

/** <p>Sets the value of the given field of the record and adds it
    to the list if field names if it was not already present.</p>
*/
- (void)setValue: (id)value forField: (NSString*)field
{
  if (![fieldNames containsObject: field])
    [fieldNames addObject: field];
  
  [recordValues setObject: value forKey: field];
}

- (void)setRecordTypes: (NSArray *)rtInfo
{
  ASSIGN(recordTypes, rtInfo);
}

- (NSArray *)recordTypes
{
  return recordTypes;
}


- (void)loadFieldValues
{
  unsigned int i;
  unsigned int sizeCount;
  NSMutableArray *fieldsArray;

  if (dbs == nil)
    return;

  fieldsArray = [NSMutableArray arrayWithCapacity: 10];
  sizeCount = 0;
  for (i = 0; i < [fieldNames count]; i++)
    {
      NSString *currField;

      currField = [fieldNames objectAtIndex:i];
      if (sizeCount + [currField length] + 2 < MAX_SOQL_SIZE)
	{
	  [fieldsArray addObject: currField];
	  sizeCount += [currField length] + 2;
	}
      else
	{
	  NSLog(@"we have %u fields", [fieldsArray count]);
	  [self loadValuesForFields: fieldsArray];
	  fieldsArray = [NSMutableArray arrayWithCapacity: 10];
	  [fieldsArray addObject: currField];
	  sizeCount = [currField length] + 2;
	}
    }
  NSLog(@"we have %u fields", [fieldsArray count]);
  [self loadValuesForFields: fieldsArray];
}

- (void)loadValuesForFields:(NSArray *)namesArray
{
  NSMutableString *statement;
  NSMutableArray *tempArray;
  DBSObject *tempObj;
  int i;

  if ([namesArray count] == 0)
    return;

  if ([self sfId] == nil)
    return;

  statement = [NSMutableString stringWithString:@"Select "];
  i = 0;
  while (i < [namesArray count])
    {
      [statement appendString: [namesArray objectAtIndex: i]];
      if (i < ([namesArray count] - 1))
	[statement appendString: @", "];
      i++;
    }
  [statement appendString: @" from "];
  [statement appendString: [self name]];
  [statement appendString: @" where id = '"];
  [statement appendString: [self sfId]];
  [statement appendString: @"'"];
  NSLog(@"query: %@", statement);
  tempArray = [NSMutableArray arrayWithCapacity: [namesArray count]];
  [dbs query :statement queryAll:YES toArray: tempArray progressMonitor:nil];
  if ([tempArray count] == 0)
    {
      [[NSException exceptionWithName:@"DBException" reason:@"Object not found." userInfo:nil] raise];
      return;
    }
  tempObj = [tempArray objectAtIndex: 0];

  /* copy the field values from the query result to the object itself */
  for (i = 0; i < [namesArray count]; i++)
    {
      NSString *fieldName;

      fieldName = [namesArray objectAtIndex: i];
      [self setValue: [tempObj valueForField: fieldName] forField: fieldName];
    }
}

@end
